framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,901 @@
+                        field.type, TypeUseLocation.FIELD);
+            }
+            for (AMethod method : aclass.methods.values()) {
+                // Return type
+                removeIgnoredAnnosFromATypeElement(
+                        method.returnType, TypeUseLocation.RETURN);
+                // Parameter type
+                for (AField param : method.parameters.values()) {
+                    removeIgnoredAnnosFromATypeElement(
+                            param.type, TypeUseLocation.PARAMETER);
+                }
+            }
+        }
+    }
+
+    /**
+     * Removes all annotations that should be ignored from an ATypeElement.
+     * (See {@link #shouldIgnore}).
+     */
+    private static void removeIgnoredAnnosFromATypeElement(ATypeElement typeEl,
+            TypeUseLocation loc) {
+        Set<Annotation> annosToRemove = new HashSet<>();
+        Set<String> annosToIgnoreForLocation = annosToIgnore.get(loc);
+        if (annosToIgnoreForLocation == null) return; // No annotations to ignore for that position.
+        for (Annotation anno : typeEl.tlAnnotationsHere) {
+            if (annosToIgnoreForLocation.contains(anno.def().name)) {
+                annosToRemove.add(anno);
+            }
+        }
+        typeEl.tlAnnotationsHere.removeAll(annosToRemove);
+
+        // Remove annotations recursively for inner types.
+        for (ATypeElement innerType : typeEl.innerTypes.values()) {
+            removeIgnoredAnnosFromATypeElement(innerType, loc);
+        }
+    }
+
+    /**
+     * Updates the parameter types of the method methodElt in the Scene of the
+     * receiverTree's enclosing class.
+     * <p>
+     * For each method parameter in methodElt:
+     *   <ul>
+     *     <li>If the Scene does not contain an annotated type for that
+     *     parameter, then the type of the respective value passed as argument
+     *     in the method call methodInvNode will be added to the parameter in
+     *     the Scene.</li>
+     *     <li>If the Scene previously contained an annotated type for that
+     *     parameter, then its new type will be the LUB between the previous
+     *     type and the type of the respective value passed as argument in the
+     *     method call.</li>
+     *   </ul>
+     * <p>
+     * @param methodInvNode the node representing a method invocation.
+     * @param receiverTree the Tree of the class that contains the method being
+     * invoked.
+     * @param methodElt the element of the method being invoked.
+     * @param atf the annotated type factory of a given type system, whose
+     * type hierarchy will be used to update the method parameters' types.
+     */
+    public static void updateInferredMethodParametersTypes(
+            MethodInvocationNode methodInvNode, Tree receiverTree,
+            ExecutableElement methodElt, AnnotatedTypeFactory atf) {
+        if (receiverTree == null) {
+            // TODO: Method called from static context.
```

Comment:
Yes, it is only for static context because of the following piece of code before calling this method in `CFAbstractTransfer`:

```
Tree receiverTree = n.getTarget().getReceiver().getTree();
if (receiverTree == null)  {
    receiverTree = analysis.getContainingClass(n.getTree());
    // receiverTree could still be null after the call above. That
    // happens when the method is called from a static context.
}
analysis.atypeFactory.getWholeProgramInference().updateInferredMethodParametersTypes(
                    n, receiverTree, method, analysis.getTypeFactory());
```